### PR TITLE
update for handle failed deployment case

### PR DESCRIPTION
**Pull request by nullfunc**
Handle FAILED deploy status

https://github.com/DefangLabs/defang/issues/491

DO NOT MERGE until https://github.com/DefangLabs/defang-mvp/pull/762 is approved

### CODE REVIEW

File: src/cmd/cli/command/commands.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ import (
 const DEFANG_PORTAL_HOST = "portal.defang.dev"
 const SERVICE_PORTAL_URL = "https://" + DEFANG_PORTAL_HOST + "/service"
 
+var ErrorFailedToReachRunningState = errors.New("failed to reach running state")
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:33]: Make these `ErrXyz` conform Go standard

File: src/cmd/cli/command/commands.go [hunk 2]
```diff
@@ -841,15 +845,20 @@ var composeUpCmd = &cobra.Command{
 			}
 		}()
 		if err := waitServiceStatus(ctx, cli.ServiceStarted, serviceInfos); err != nil && !errors.Is(err, context.Canceled) {
-			if !errors.Is(err, cli.ErrDryRun) && !errors.As(err, new(cliClient.ErrNotImplemented)) {
+			if errors.Is(err, ErrorDeploymentFailed) {
+				term.Warn("Deployment FAILED. Service(s) not running.")
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:849]: `return err` here. This would show the error message and exit the CLI with a non-zero exit code = win win

**Review by lionello — CHANGES_REQUESTED** [review]

**Comment by lionello** [comment]
I want us to fix up the existing `serviceInfo.Status` field to match the Status returned by `Subscribe`. I think both of these should be the same and should be an `enum`. The `serviceInfo.Status` field will also contain a free form string, which we'd have to create a new field for (or rather, put the string in the existing "Status" field but add the new enum field.)

File: src/protos/io/defang/v1/fabric.proto [hunk 3]
```diff
@@ -117,6 +117,24 @@ message UploadURLRequest { string digest = 1; }
 
 message UploadURLResponse { string url = 1; }
 
+
+enum ServiceState {
+  UNKNOWN = 0;
+
+  // Build states (initial state for build)
+  BUILD_PENDING = 1; // BUILD_QUEUED -> BUILD_ACTIVATING
```
Comment (lionello) [inline src/protos/io/defang/v1/fabric.proto:125]: I'm not sure why these aren't simply 1:1 the existing ones? Seems not too much gain for lot of confusion

File: src/protos/io/defang/v1/fabric.proto [hunk 4]
```diff
@@ -117,6 +117,24 @@ message UploadURLRequest { string digest = 1; }
...
+  BUILDING = 2; // BUILD_RUNNING
+  CREATED = 3; // BUILD_DEACTIVATING (done = true)
+ 
+  // Update states (initial state for existing image)
+  UPDATE_PENDING = 4; //  UPDATE_QUEUE
+
+  // Deployment states
+  DEPLOYING = 5;
```
Comment (lionello) [inline src/protos/io/defang/v1/fabric.proto:133]: especially these: PENDING, COMPLETED, FAILED are pretty clear, why change it up?

File: src/protos/io/defang/v1/fabric.proto [hunk 5]
```diff
@@ -117,6 +117,24 @@ message UploadURLRequest { string digest = 1; }
...
+enum ServiceState {
+  UNKNOWN = 0;
+
+  // Build states (initial state for build)
+  BUILD_PENDING = 1; // BUILD_QUEUED -> BUILD_ACTIVATING
+  BUILDING = 2; // BUILD_RUNNING
+  CREATED = 3; // BUILD_DEACTIVATING (done = true)
+ 
+  // Update states (initial state for existing image)
+  UPDATE_PENDING = 4; //  UPDATE_QUEUE
```
Comment (lionello) [inline src/protos/io/defang/v1/fabric.proto:130]: again, why the rename?

**Review by lionello — CHANGES_REQUESTED** [review]

File: src/protos/io/defang/v1/fabric.proto (hunk 5 above)
Comment (nullfunc) [reply src/protos/io/defang/v1/fabric.proto:130]: Will change back.

File: src/protos/io/defang/v1/fabric.proto (hunk 3 above)
Comment (nullfunc) [reply src/protos/io/defang/v1/fabric.proto:125]: I was trying to map them to something platform agnostic but probably a little early to be doing that.

File: src/protos/io/defang/v1/fabric.proto [hunk 6]
```diff
@@ -117,6 +117,28 @@ message UploadURLRequest { string digest = 1; }
 
 message UploadURLResponse { string url = 1; }
 
+
+enum ServiceState {
+  SERVICE_STATE_UNSPECIFIED = 0;
```
Comment (lionello) [inline src/protos/io/defang/v1/fabric.proto:122]: just `UNSPECIFIED`

**Review by lionello — APPROVED** [review]

**Pull request merged**